src/etc/inc/interfaces.inc

```diff
@@ -2401,14 +2401,17 @@ function interface_configure($verbose = false, $interface = 'wan', $reload = fal
      * which triggers the interface config again, which attempts to
      * spoof the MAC again which cycles the link again...
      */
-    if (!empty($wancfg['spoofmac']) && strcasecmp($wancfg['spoofmac'], get_interface_mac($realhwif))) {
-        mwexecf('/sbin/ifconfig %s link %s', array($realhwif, $wancfg['spoofmac']));
+    if (!empty($wancfg['spoofmac']) && strcasecmp($wancfg['spoofmac'], get_interface_mac($realif))) {
+        mwexecf('/sbin/ifconfig %s link %s', array($realif, $wancfg['spoofmac']));
 
-        /* All vlans need to spoof their parent mac address, too. */
-        if (isset($config['vlans']['vlan'])) {
+        /* A MAC address change on a parent interface resets all VLANs to the parent MAC, so we have to restore spoofed VLAN MACs. */
+        if (!stristr($realif, "_vlan") && isset($config['vlans']['vlan'])) {
             foreach ($config['vlans']['vlan'] as $vlan) {
                 if ($vlan['if'] == $realhwif) {
-                    mwexecf('/sbin/ifconfig %s link %s', array($vlan['vlanif'], $wancfg['spoofmac']));
+                    $friendly_vlanif = convert_real_interface_to_friendly_interface_name($vlan['vlanif']);
+                    if (!empty($friendly_vlanif) && !empty($config['interfaces'][$friendly_vlanif]['spoofmac'])) {
+                        mwexecf('/sbin/ifconfig %s link %s', array($vlan['vlanif'], $config['interfaces'][$friendly_vlanif]['spoofmac']));
+                    }
```

Comment:
@maurice-w shouldn't there be an `else` here setting the macaddr to the interface address when not overwritten by the vlan itself? at a first glance I would expect this can break existing setups depending on the mac address being equal on both vlan and interface.

Reply:
@AdSchellevis Setting the MAC address on a physical interface trickles down to the VLANs automatically. This happens on the OS level, we don't have to do it manually. Why did the original code do it anyway? I have no idea.

So the default behavior doesn't change: Configuring a spoofed MAC on a parent is applied to all of its VLANs, too (unless they have their own spoofed MAC).

Reply:
@maurice-w I guess it's in the order of things, if you set it, apply and remove it, it will stick to the old value...

Reply:
@AdSchellevis I'm pretty sure removing a spoofed MAC has always required a reboot. We don't do anything if `$wancfg['spoofmac']` is empty so the old spoofed MAC sticks until the interface gets reinitialized on boot.

Reply:
@maurice-w I mean the other way around, set a mac on the vlan, apply and remove it.  In case the parent is reset again, we can't know the old value indeed, but if the vlan changes and the parent is left untouched one might expect after a save it will drop back to the interface mac.

Reply:
@AdSchellevis, please help me out here, I don't get it. :-)
- This is no longer about breaking existing setups, is it? _"set a mac on the vlan, apply and remove it"_ never worked without a reboot. You're requesting new / improved functionality?
- Since we could only make this work for VLANs, we would shift from "always requires a reboot" to "sometimes works, but sometimes requires a reboot, depending on the type of interface". Do we want that?
- An `else` here would do nothing because we don't get this far if `$wancfg['spoofmac']` is empty. We would have to restructure the entire flow. I intended this change to be limited, but if desired, I guess I can do that.

Reply:
@maurice-w It's about a bit about "breaking", with vlan's supporting custom mac addresses, the remove operation isn't functional anymore with this change. It's easy to explain that you do need a reboot to reset back to factory address, a little less obvious that removing a vlan mac address doesn't bring it back to it's parent one like it previously did. 

I guess my single point is that now you can end up in a state where some mac addresses of the interface are different but configured equal, where in the past they where always the same. I know we can't retrieve the original value if the parent is reconfigured, no problem, I'm just very sure one day someone reports an issue about "not all vlan's are equal, but configured the same" which will be difficult to track.

If the whole intend it to clean the code and always set the mac address when configured and don't try to sync vlan's, this probably isn't the right fix. (removing all vlan code here likely has the same effect)

Reply:
> removing a vlan mac address doesn't bring it back to it's parent one **like it previously did**

Sorry, but no, it never did. Not without a reboot. That's my point. Please try it. Set a spoofed MAC on a VLAN, apply, remove the spoofed MAC, apply again. The spoofed MAC sticks until you reboot or apply a different spoofed MAC. The patch doesn't change this at all. And a reboot still brings the VLANs back to the parent MAC.

I'm happy to talk about functional improvements, but would like to get the "breaks existing setups" concern out of the way first.

Reply:
I'll keep the PR open to discuss internally later, I see your point, just don't have a lot of time to dig into this now.

Reply:
Okay. Just to clarify: The patch is meant to fix a specific issue without breaking anything. And I'm convinced it does. It's not meant to solve _all_ MAC address spoofing related issues once and forever. Continual improvement.

Reply:
yes, that's clear, before pulling it in we do need to assess the impact though. The connectivity between the interface and the associated vlan from the system part is just a bit odd. Maybe that changed somewhere over time, I don't know.

Reply:
Came here by a roundabout method trying to work out why the LAN interface on PC was getting RAs from different VLANs, Checked my switches, no problem there, checked the firewall, removing 134s from filter.lib.inc to narrow it down; problem is definitely on the firewall. So Checked all my rules, they seem fine, checked whether there were any tunables that might be to blame.,, So I set the interface to Promiscuous and installed this patch.. problem solved. Thanks @maurice-w, spent half a day trying to fix  this as I'm pretty certain it never used to happen, but hey, everything appeared to be working fine anyway so I wasn't really looking.

Reply:
Great to hear @marjohn56! Unfortunately, adding a reliable and user-friendly way to enable promiscuous mode seems harder than expected (see my last comment below). Would you by any chance be interested in helping with that? :-)

Reply:
Happy to, though my time is sadly limited by real work, only had a bit of time to play today as what was supposed to happen didn't and it was too late to arrange to do anything else!

There's nothing particularly complex on my system. It's a simple Qotom system where I am using one NIC and pushing three VLANs down it, there is no untagged LAN per say, just the three VLANs; connects straight to one of the trunks on my first switch and out to the rest of the switches from there. I'll comment on your last comment now.